csrc/id_model/id_graphs.cpp

```diff
@@ -0,0 +1,1768 @@
+      p2c_ca_permissive_maps;
+
+  // All producer ids within (including dependencies of) inlined leaf domains,
+  // used for deterministic order
+  VectorOfUniqueEntries<IterDomain*> ordered_p_ca_ids;
+
+  std::unordered_map<IterDomain*, VectorOfUniqueEntries<IterDomain*>>
+      p2c_root_broadcast_resolution_map;
+};
+
+// Returns the root producer iteration domains that are resolved by provided
+// consumer
+std::unordered_map<IterDomain*, IterDomain*> resolvedRootBroadcasts(
+    TensorView* producer,
+    TensorView* consumer) {
+  auto p2c_map =
+      PairwiseRootDomainMap(producer, consumer)
+          .mapProducerToConsumer(producer->domain(), consumer->domain());
+
+  std::unordered_map<IterDomain*, IterDomain*> resolved_bcast_map;
+  for (const auto& kv : p2c_map) {
+    auto p_id = kv.first;
+    // Ignore non-broadcast dims
+    if (!p_id->isBroadcast()) {
+      continue;
+    }
+    auto c_id = kv.second;
+    // If the consumer ID is a reduction (i.e., a trivial
+    // reduction), do not consider it's concretized.
+    if (c_id->isBroadcast() || c_id->isReduction()) {
+      continue;
+    }
+
+    resolved_bcast_map[p_id] = c_id;
+  }
+  return resolved_bcast_map;
+}
+
+StatefulLoweringInfo buildInfo(
+    const std::vector<Expr*>& exprs,
+    const IdGraph& exact_graph,
+    const IdGraph& permissive_graph) {
+  StatefulLoweringInfo info;
+  // Grab inlining relationships
+  for (auto expr : exprs) {
+    for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+      auto producer_root = producer->getMaybeRFactorDomain();
+      auto producer_domain = producer->domain()->leaf();
+
+      // Grab all iteration domains in producer that its compute at iter domains
+      // depend on.
+      VectorOfUniqueEntries<IterDomain*> all_producer_ca_deps;
+      {
+        auto ca_dep_vals = DependencyCheck::getAllValsBetween(
+            {producer_root.begin(), producer_root.end()},
+            {producer_domain.begin(),
+             producer_domain.begin() + producer->getComputeAtPosition()});
+        auto ca_deps_filter = ir_utils::filterByType<IterDomain>(ca_dep_vals);
+
+        all_producer_ca_deps.insert(
+            ca_deps_filter.begin(), ca_deps_filter.end());
+      }
+
+      info.ordered_p_ca_ids.pushBack(all_producer_ca_deps);
+
+      for (auto consumer :
+           ir_utils::filterByType<TensorView>(expr->outputs())) {
+        auto resolved_bcast_map = resolvedRootBroadcasts(producer, consumer);
+        for (auto entry : resolved_bcast_map) {
+          info.p2c_root_broadcast_resolution_map[entry.first].pushBack(
+              entry.second);
+          for (auto other_exact_bcast : *exact_graph.toGroup(entry.first)) {
+            if (all_producer_ca_deps.has(other_exact_bcast)) {
+              info.p2c_root_broadcast_resolution_map[other_exact_bcast]
+                  .pushBack(entry.second);
+            }
+          }
+        }
```

Comment:
The broadcast resolution map is set for a pair of broadcast producer and non-broadcast consumer domains. It's also set for the domains that are exactly mapped with the broadcast producer domain *only when those mapped producers are included in `all_producer_ca_deps`*. Why the original producer broadcast domain is not checked if it's included in `all_producer_ca_deps`? I.e., why we shouldn't do this:

```suggestion
      for (auto consumer :
           ir_utils::filterByType<TensorView>(expr->outputs())) {
        auto resolved_bcast_map = resolvedRootBroadcasts(producer, consumer);
        for (auto entry : resolved_bcast_map) {
          if (all_producer_ca_deps.has(entry.first)) {
            info.p2c_root_broadcast_resolution_map[entry.first].pushBack(
                entry.second);
          }
          for (auto other_exact_bcast : *exact_graph.toGroup(entry.first)) {
            if (all_producer_ca_deps.has(other_exact_bcast)) {
              info.p2c_root_broadcast_resolution_map[other_exact_bcast]
                  .pushBack(entry.second);
            }
          }
        }
```

Reply:
The logic doesn’t make a lot of sense to me. I would think it should be:
```.cpp
      for (auto consumer :
           ir_utils::filterByType<TensorView>(expr->outputs())) {
        auto resolved_bcast_map = resolvedRootBroadcasts(producer, consumer);
        for (auto entry : resolved_bcast_map) {
          if (all_producer_ca_deps.has(entry.first)) {
            info.p2c_root_broadcast_resolution_map[entry.first].pushBack(
                entry.second);
            for (auto other_exact_bcast : *exact_graph.toGroup(entry.first)) {
                info.p2c_root_broadcast_resolution_map[other_exact_bcast]
                    .pushBack(entry.second);
            }
          }
        }
```

Reply:
Or maybe there shouldn’t even be that inner most for loop.